### ISSUE TITLE
Upgrade various dependencies 

### ISSUE THREAD

**Issue by marceljay** [kumodao/kumo-protocol#390]
The project is a fork of Liquity, which seemingly happened in January 2022, and it's not clear (at least to me), how outdated dependencies have been dealt with. Since this is a large mono-repo with many interdependent packages, the complexity of maintaining all these dependencies is even harder. 

Recently a lot of issues occurred (subgraph libs, synpress, hardhat/solidity) which are related to outdated dependencies. 
Upgrading these packages often comes with high risk of breaking things, which is why a strategy and care are necessary to gradually move to a more up-to-date dependency tree. 
 
This issue should discuss how we want to upgrade dependencies gradually. 

cc: @asimpk @arthurka-o @udosson

**Comment by udosson** [comment]
@marceljay - thanks for creating this issue.
What do you generally think of moving from a mono-repo to separated repos? 
e.g. smart contracts, SDK, font-end?

**Comment by asimpk** [comment]
Hi, I think we should stick with Monorepos, Although these have pros and cons, also is recommended by experts. The important thing is how well we can manage it.

**Comment by asimpk** [comment]
> This issue should discuss how we want to upgrade dependencies gradually.
We can do it by one package at a time and then if any issue arises then should be fixed before moving further. Mostly I observed issues in Jest testing files (only related to typescript type) and also in lib-ethers or providers package (mostly related to ethers typescript package). 
This might be helpful https://dev.to/ttoss/maintaining-a-monorepo-strategies-for-managing-dependencies-and-updating-packages-3bfl

For example: I tried from the root dir package.json
`npx npm-check-updates -u`

Output:
 @microsoft/api-documenter        ^7.17.14  →  ^7.22.33
 fs-extra                                          ^9.1.0  →   ^11.1.1
 lint-staged                                     ^14.0.0  →   ^14.0.1
 prettier                                          ^3.0.1  →    ^3.0.3
 semantic-release                           ^19.0.3  →   ^21.1.1
 semantic-release-monorepo         ^7.0.4  →    ^7.0.5
 
 After running yarn on this, I got the error 
 error semantic-release@21.1.1: The engine "node" is incompatible with this module. Expected version ">=18". Got "16.16.0"
 
 So for semantic-release package, I switched back to ^19.0.3 and tried again and this time no error.

**Comment by marceljay** [comment]
> After running yarn on this, I got the error error semantic-release@21.1.1: The engine "node" is incompatible with this module. Expected version ">=18". Got "16.16.0"
> 
> So for semantic-release package, I switched back to ^19.0.3 and tried again and this time no error.

Great approach!

**Comment by asimpk** [comment]
Steps that I followed to upgrade the node and dependencies. 
Specified latest recommended node LTS version in root package.json "node": "^18.17.1"
run npx npm-check-updates -u for every package.json to specify all the packages to the latest versions.
Iteration of yarn install
The first Iteration I found https://www.npmjs.com/package/@apollo/federation fails the installation, it only works with node less than 17. 
Instead of blindly trying to fix this I skip the lib-subgraph, and subgraph packages (I know How can try to fix them later).
Next the code breaks everywhere, I first fixed the SDK packages (lib-base, lib-ethers, lib-react, providers) and at last dev-frontend.
These fixes are not ready to merge (I used a more generic approach while fixing, we should be more specific while working with typescript unless it is necessary). 
I tried to upgrade yarn to version 3 but I got two issues migrations and dependencies. So I had two options either I should first upgrade to yarn3 and spend time to make migration work properly or solve the dependencies issues first. So I decided to upgrade the dependencies first although I understood it would take longer everything is set, and at least we can get the real picture of issues no matter whether we use the yarn v1 or v3.

**Comment by asimpk** [comment]
[](https://kumo-space.slack.com/archives/C049S585CT0/p1694972970310889?thread_ts=1694709956.541109&cid=C049S585CT0)
Yeah I'm upgrading the node version to the latest "^18.17.1"

[](https://kumo-space.slack.com/archives/C049S585CT0/p1694973711107329?thread_ts=1694709956.541109&cid=C049S585CT0)
It depends on how we specify our dependencies, in the case of liquity they added a wide range of node versions but If you see liquity tightly fixed the dependencies versions in all the packages. If you make any changes to those dependencies it starts creating issues and breaking, so it means that the fixed version works for a wide range they specified. 
In our case, after upgrading the node every package that will be in use will be up to date and should work. If we add some node versions range then we need to select for each dependency which should work for all so that will be not sure that these dependencies are the latest as maybe any latest version may not work with node 14 and 16 but work with 17 or 18 only.

**Comment by marceljay** [comment]
> The first Iteration I found https://www.npmjs.com/package/@apollo/federation fails the installation, it only works with node less than 17.
> Instead of blindly trying to fix this I skip the lib-subgraph, and subgraph packages (I know How can try to fix them later).
> Next the code breaks everywhere,
@asimpk 
1) I don't really understand how you did this... so you saw installation issues and then skipped packages during installation, meaning you did only do a partial install? Can you log what commands you have used or post a long of the whole terminal output in a Gist or something?

2) "Next the code breaks everywhere..." during what? `run-s prepare:*`?

**Comment by marceljay** [comment]
> > The first Iteration I found https://www.npmjs.com/package/@apollo/federation fails the installation, it only works with node less than 17.
> > Instead of blindly trying to fix this I skip the lib-subgraph, and subgraph packages (I know How can try to fix them later).
> > Next the code breaks everywhere,
> > @asimpk
> 
> 1. I don't really understand how you did this... so you saw installation issues and then skipped packages during installation, meaning you did only do a partial install? Can you log what commands you have used or post a long of the whole terminal output in a Gist or something?
> 2. "Next the code breaks everywhere..." during what? `run-s prepare:*`?

**Comment by marceljay** [comment]
And I'm not sure how much time you've spent on this, but my intuition is that with Yarn v3 we can fix the dependency issues much faster.

**Comment by asimpk** [comment]
> > The first Iteration I found https://www.npmjs.com/package/@apollo/federation fails the installation, it only works with node less than 17.
> > Instead of blindly trying to fix this I skip the lib-subgraph, and subgraph packages (I know How can try to fix them later).
> > Next the code breaks everywhere,
> > @asimpk
> 
> 1. I don't really understand how you did this... so you saw installation issues and then skipped packages during installation, meaning you did only do a partial install? Can you log what commands you have used or post a long of the whole terminal output in a Gist or something?
The first step was to make changes in the packag.json dependencies with their latest version, next step was to install these dependencies.  As I upgraded the Node version to 18.x.x then command 
yarn install 
fails with this error 
Fetching packages...
error @apollo/federation@0.27.0: The engine "node" is incompatible with this module. Expected version ">=12.13.0 <17.0". Got "18.17.1"
error Found incompatible module.

I tracked the error and found subgraph package is the source, so for the time being I skipped these packages which are not compatible with node 18 or even 17 and we are not currently directly using these packages. But It doesn't my intent is to completely remove or skip these, I just wanted to make sure all the dependencies work or not or which packages are compatible or not.

> 2. "Next the code breaks everywhere..." during what? `run-s prepare:*`?
Sorry for the bad word "breaks everywhere", it actually not breaks but a few errors in every package of workplace and yes while running this command `yarn prepare`

**Comment by asimpk** [comment]
> And I'm not sure how much time you've spent on this, but my intuition is that with Yarn v3 we can fix the dependency issues much faster.

Yeah, you are right, I'm migrating to Yarn v3, as I already started before and fixing issues.

**Comment by marceljay** [comment]
thanks for clarifying, but still not sure what you mean by "skip"? 

> I tracked the error and found subgraph package is the source, so for the time being I **skipped these packages** which are not compatible with node 18 or even 17 and we are not currently directly using these packages.

Afaik, packages can be removed but how can they be skipped? 
And what makes you think we don't need them? Could you list the packages?

**Comment by asimpk** [comment]
> thanks for clarifying, but still not sure what you mean by "skip"?
>  As you said it's important to follow incremental progress and that was also in my mind when I started this upgradation 
    task. Skip means not including these packages in the first one or two PRs related to the updating process. Otherwise, PR will 
    be large and so many changes to review.
> > I tracked the error and found subgraph package is the source, so for the time being I **skipped these packages** which are not compatible with node 18 or even 17 and we are not currently directly using these packages.
> Yes one of the nested dependencies of the subgraph package is not compatible which then breaks the subgraph `yarn 
   prepare` command fails during installation or building packages. which also makes fail lib-subgraph and fuzzier packages.  
> Afaik, packages can be removed but how can they be skipped? And what makes you think we don't need them? Could you list the packages?
Can you please check this change, I listed only those packages that are successful to build. 
https://github.com/kumodao/kumo-protocol/pull/405/commits/1bbfa31e6299d0a44a516a419c0d107c1215382b#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R6
I did not list these three packages
https://github.com/kumodao/kumo-protocol/tree/dev/packages/fuzzer
https://github.com/kumodao/kumo-protocol/tree/dev/packages/lib-subgraph
https://github.com/kumodao/kumo-protocol/tree/dev/packages/subgraph

>> And what makes you think we don't need them?
No, it doesn't mean that we don't need these above packages, what I was trying to explain we are not using these with frontend, and our complete functionality is not dependent on these packages. Also, I understand where was the problem and how we can fix it in next PRs.

**Comment by asimpk** [comment]
> Do you know how the subgraph package is being used or do you have any docs you point me to?

We have only this information and I followed this documentation and after resolving configuration issues ran the subgraph node locally and accessed the client interface in the browser, I shared the related stuff in Slack but unfortunately, we can't access that now (also that was not a specific task assigned to me and not the part of our MVP, I did only to understand what Subgraph is and how it works etc).
